clang/include/clang/Basic/Attr.td

```diff
@@ -3061,6 +3061,7 @@ def WorkGroupSizeHint :  InheritableAttr {
     }
   }];
   let Documentation = [WorkGroupSizeHintAttrDocs];
+  let SupportsNonconformingLambdaSyntax = 1;
```

Comment:
Oh, I see now from the summary!

The SYCL specification was wrong to suggest this syntactic location in the first place (it's not valid C++ code and is not a conforming extension to C++ either). So we can either follow a known-broken specification or we can assume the spec has a defect report and support the correct syntactic location. I fall on the side of not following the broken spec, but I won't block the review on it either.

(To be clear, `SupportsNonconformingLambdaSyntax` will never be upstreamed into Clang, so gaining further reliance on it should keep that mind.)

Reply:
> Oh, I see now from the summary!
> 
> The SYCL specification was wrong to suggest this syntactic location in the first place (it's not valid C++ code and is not a conforming extension to C++ either). So we can either follow a known-broken specification or we can assume the spec has a defect report and support the correct syntactic location. I fall on the side of not following the broken spec, but I won't block the review on it either.
> 
> (To be clear, `SupportsNonconformingLambdaSyntax` will never be upstreamed into Clang, so gaining further reliance on it should keep that mind.)

Yes, C++ does not allow attributes on lambda functions (yet ?) which confused me while checking the SYCL specification on this point. I agree on sticking to 100% conformant C++. And so, I tried rewriting some code to use functors instead of lambdas in order to use the attributes in the conformant places. But something breaks during the code generation and DPC++ generates terrible code on CUDA. Everything goes to the stack while with the lambda version uses no stack. The code is on GitHub if you believe it's worth looking into it

Reply:
They're allowed in C++23, which Clang 13 and later handles as a conforming extension in earlier language modes. See: https://godbolt.org/z/W39K31EYa